---
title: "child_manuscript_tables"
author: "Mohamed_Shoeb"
date: "10/20/2021"
output: html_document
---

Create tables output directory
```{r}
dir.create(here("figures_tables"))
dir.create(here("supp_tables"))
```

# Main figures tables
```{r figure_3}
#Figure 3 B,C,D
fig3_bcd <- srt_obj@reductions$umap@cell.embeddings %>%
  cbind(srt_obj@meta.data)%>%
  as.data.frame() %>%
  mutate(predicted.ref1_cell_type = str_to_sentence(predicted.ref1_cell_type),
         predicted.ref1_cell_type = str_replace(predicted.ref1_cell_type, "-", " "),
         predicted.ref1_cell_type = str_replace(predicted.ref1_cell_type, " t ", " T "),
         predicted.ref1_cell_type = str_replace(predicted.ref1_cell_type, " b ", " B "),
         predicted.ref2_genotype = factor(predicted.ref2_genotype,
                                          levels = c("Tg(mpx:EGFP)",
                                                     "Tg(cd41:EGFP)",
                                                     "Tg(runx1:mCherry)",
                                                     "Tg(cd4:mCherry)",
                                                     "Tg(tal1:EGFP)",
                                                     "Tg(gata1a:GFP)",
                                                     "Tg(mfap4:tdTomato)")),
         predicted.ref1_cell_type = ifelse(predicted.ref1_cell_type =="Neutrophil and myeloid cell", "Neutrophil", "Others"),
         predicted.ref2_genotype = ifelse(predicted.ref2_genotype =="Tg(mpx:EGFP)", "Neutrophil", "Others")
         ) %>% 
  dplyr::select(UMAP_1,
                UMAP_2,
                citrin_state,
                predicted.ref1_cell_type,
                predicted.ref2_genotype) %>% 
    rownames_to_column("cell_id") %>%
  dplyr::select(cell_id, everything())

write_csv(fig3_bcd,file.path("figures_tables", "fig3_bcd.csv"))

#Figure 3 E,F,G
fig3_efg <- ss_df_traj %>% 
  mutate(citrin_state = factor(citrin_state, levels = c("WKM", "HI", "INT", "NO")),
         Sort = ifelse(citrin_state == "WKM","all kidney marrow","Neutrophil sort"),
         fish = NULL,
         pool = NULL,
         tissue = NULL,
         mmp9 = srt_traj@assays$SCT@data["mmp9",cell_id],
         lyz = srt_traj@assays$SCT@data["lyz",cell_id])

write_csv(fig3_efg,file.path("figures_tables", "fig3_efg.csv"))

```

```{r Figure_4}
#Figure 4A
##top annotation
fig4_a_top_annot <- cell_avg_hc_labels  %>%
  stack %>%
  dplyr::rename(pseudotime_bin = "ind",
                phase_annotation = "values") %>%
  dplyr::select(pseudotime_bin, phase_annotation)

write_csv(fig4_a_top_annot,file.path("figures_tables", "fig4_a_top_annot.csv"))

##heatmap values
fig4_a_heatmap<- aggr_avg_traj_scaled %>% 
  rownames_to_column("gene_symbol") %>% 
  dplyr::select(gene_symbol, everything()) 
write_csv(fig4_a_heatmap,file.path("figures_tables", "fig4_a_heatmap.csv"))

##left annotations
fig4_a_left_annot <- gene_clusters_avg_hc%>% 
  rownames_to_column("gene_symbol") %>% 
  dplyr::select(gene_symbol, everything())
write_csv(fig4_a_left_annot,file.path("figures_tables", "fig4_a_left_annot.csv"))

#Figure 4B
fig4_b <- tall_df %>%
  pivot_longer(-c(gene_symbol,module)) %>%
   group_by(name, module) %>%
    summarize(avg_scaled_expression = mean(value)) %>%
     ungroup() %>%
     mutate(module = factor(module, levels = paste0("module_", 3:1), labels = c(module_3 ="M3",module_2 ="M2",module_1 ="M1")),
            phase = factor(name, labels = c(phase_1 ="P1",phase_2 ="P2",phase_3 ="P3",phase_4 ="P4")))

write_csv(fig4_b,file.path("figures_tables", "fig4_b.csv"))

#Figure 4 C
# top fdr; x axis GeneRatio; top 5
fig4_c <- lapply(hypeR_res_msig, function(module){
 module %>%
  filter(fdr<0.1)%>%
  #group_by(group) %>% 
  #slice_min(fdr, n = 5,with_ties = FALSE) %>% 
  #ungroup() %>% 
  mutate(GeneRatio = overlap/geneset) %>% 
  arrange((fdr)) %>%
  head(6)%>%  #top hits
  arrange((GeneRatio))
  }) %>% 
  bind_rows(.id = "module") %>%
  mutate(label = label %>% str_replace_all("_|GOBP|HALLMARK|REACTOME|KEGG"," "),
         label = factor(label, levels = unique(label)),
          module = factor(module,labels = c(module_1 ="M1",module_2 ="M2",module_3 ="M3"))) %>% 
  dplyr::select(-hits ) %>% 
  arrange(module, desc(GeneRatio))

write_csv(fig4_c,file.path("figures_tables", "fig4_c.csv"))

#Figure 4 D
groups <- c("oxidative_stress", "inflammation")
l <- lapply(names(group_expr_matrix_lst)[-7], function(gene_group){
  #expression matrix to the corresponding group
  aggr_avg_traj_stefanie <- group_expr_matrix_lst[[gene_group]]
  stefanie_genes <- rownames(aggr_avg_traj_stefanie)
  #scaled matrix
  aggr_avg_traj_scaled_stefanie <- aggr_avg_traj_stefanie %>% 
    t() %>% 
    scale() %>% 
    t() %>% 
    as.data.frame() %>% 
    mutate_all(~ ifelse(abs(.)>2.5, 2.5* ./abs(.),.))
  
  #Available genes and their funcion labels
  o <- seriate(aggr_avg_traj_scaled_stefanie[stefanie_genes,], method = "PCA")
  o <- if(gene_group %in% groups){
    get_order(o, 1)
    }else{
      rev(get_order(o, 1))
    }
  stefanie_genes <- stefanie_genes[o ] 
 
  #
  labels_row_sig <- intersect(stefanie_genes, diff_traj$gene_symbol)
  aggr_avg_traj_scaled_stefanie[labels_row_sig,] %>%
    rownames_to_column("gene_symbol")
  })
names(l) <-names(group_expr_matrix_lst)[-7]

figure4_d <- l %>%
  bind_rows(.id = "gene_group")

write_csv(figure4_d,file.path("figures_tables", "fig4_d.csv"))
```

```{r}
#Figure 5a
figure5_a <- lapply("regulators", function(gene_group){
  #expression matrix to the corresponding group
  aggr_avg_traj_stefanie <- group_expr_matrix_lst[[gene_group]]
  stefanie_genes <- rownames(aggr_avg_traj_stefanie)
  #scaled matrix
  aggr_avg_traj_scaled_stefanie <- aggr_avg_traj_stefanie %>% 
    t() %>% 
    scale() %>% 
    t() %>% 
    as.data.frame() %>% 
    mutate_all(~ ifelse(abs(.)>2.5, 2.5* ./abs(.),.))
  
  #Available genes and their funcion labels
  o <- seriate(aggr_avg_traj_scaled_stefanie[stefanie_genes,], method = "PCA")
  o <- if(gene_group %in% groups){
    get_order(o, 1)
    }else{
      rev(get_order(o, 1))
    }
  stefanie_genes <- stefanie_genes[o ] 
 
  #
  labels_row_sig <- intersect(stefanie_genes, diff_traj$gene_symbol)
  aggr_avg_traj_scaled_stefanie[labels_row_sig,] %>%
    rownames_to_column("gene_symbol") %>% 
    mutate(gene_group = gene_group) %>% 
    dplyr::select(gene_group, gene_symbol, everything())
  })
figure5_a <- figure5_a[[1]]

write_csv(figure5_a,file.path("figures_tables", "fig5_a.csv"))

#Figure 5b
figure5_b <- all_tf_weight_module_df %>%
  filter(regulatoryGene %in% reg_dunn_top3$regulatoryGene) %>%
  mutate(regulatoryGene = factor(regulatoryGene, levels = rev(reg_dunn_top3$regulatoryGene)),
         dist = NULL) %>% 
  dplyr::rename("Module specificity" = "weight")

write_csv(figure5_b,file.path("figures_tables", "fig5_b.csv"))

#Figure 5c
names(assoc_reg_ls) <- gene_order[gene_order%in%names(target_min_tf)]
figure5_c <- assoc_reg_ls[c("ybx1", "cebpb")] %>%
  bind_rows() %>% 
  dplyr::rename("Normalized expression" = "value",
                "Pseudotime" = "t")

write_csv(figure5_c,file.path("figures_tables", "fig5_c.csv"))
```

```{r}
#Figure 6 a
figure6_a <- norm_lst$quants_all %>%
  as.data.frame() %>%
  rownames_to_column("gene_symbol") %>%
  mutate(module=gene_clusters_avg_hc[gene_symbol,"module"]) %>%
  dplyr::select(gene_symbol, module, everything())

write_csv(figure6_a,file.path("figures_tables", "fig6_a.csv"))

#Figure 6 B, C, D, E
figure6_bcde <- cbind(module=gene_clusters_avg_hc[gene_order_ortho,"module"],
                                 ccf_all_mean_species[gene_order_ortho,] %>% as.data.frame %>% rename_with(~paste0("cc_mean_", .x)),
                                 (ls_aggr_expr_smooth%>%
                                   do.call(cbind,.))[gene_order_ortho,]) %>%
  rownames_to_column("gene_symbol") %>% 
  as.data.frame()

write_csv(figure6_bcde,file.path("figures_tables", "fig6_bcde.csv"))

#Figure 6 F
#supplementary_table_7: cross-correlation results
figure6_f <- ccf_all_mean_species %>% 
  rownames_to_column("gene_symbol")

write_csv(figure6_f,file.path("figures_tables", "fig6_f.csv"))

```

```{r}
#Figure 7 A
figure7_a <- nboma_norm_out[nboma_ordered_samples, ] %>%
  as.data.frame() %>% 
  rownames_to_column("sample_id")

write_csv(figure7_a,file.path("figures_tables", "fig7_a.csv"))

```

# Supplementary figures tables

```{r}
#Supp figure 3  B,C,D,E,F,G,H
figure3supp_bcdefghij <-  srt_obj@reductions$umap@cell.embeddings %>%
  as.data.frame() %>% 
  rownames_to_column("cell_id") %>% 
  cbind(srt_obj@meta.data %>% dplyr::select(-cell_id)) %>% 
  as.data.frame() %>% 
  mutate(neutrophils = ifelse(cell_id %in% ss_df_traj$cell_id, "Neutrophils", "else"),
       cluster = ifelse(res_0.05 %in% c(0,1),"Major cluster","Other clusters")) %>% 
  dplyr::select(cell_id, UMAP_1,UMAP_2,fish, pool, batch, Phase, percent_mt, nCount_RNA, citrin_state, neutrophils, cluster)


write_csv(figure3supp_bcdefghij,file.path("figures_tables", "fig3supp_bcdefghij.csv"))
#
figure3supp_kl <- srt_obj@reductions$umap@cell.embeddings %>%
  cbind(srt_obj@meta.data) %>% 
  as.data.frame() %>% 
  dplyr::select(Phase, res_0.05) %>% 
  rownames_to_column("cell_id")
write_csv(figure3supp_kl,file.path("figures_tables", "fig3supp_kl.csv"))
#
figure3supp_m <- hypeR_res_msig %>% rownames_to_column("cell_id")
write_csv(figure3supp_m,file.path("figures_tables", "fig3supp_m.csv"))
```

```{r}
#Supp figure 4 A-h
fig4supp_ah <- tibble(cell_id = colnames(sce),
         cluster = SingleCellExperiment::colLabels(sce),
         slingshot_score = setNames(ss_df_traj$pseudotime,
                              ss_df_traj$cell_id)[colnames(sce)] %>% as.numeric(),
         tscan_score = as.numeric(pseud_time),
         monocle_score = cds@principal_graph_aux@listData[["UMAP"]][["pseudotime"]][cell_id],
         first_component = setNames(ss_df_traj$harmony_1,
                              ss_df_traj$cell_id)[colnames(sce)]
           ) %>% 
  arrange(slingshot_score) 

write_csv(fig4supp_ah,file.path("figures_tables", "fig4supp_a-h.csv"))
```

```{r}
#Supp figure 5 
figure5supp <- all_tf_weight_module_df %>%
  mutate(regulatoryGene = factor(regulatoryGene, levels = (reg_dunn_all$regulatoryGene)),
         dist = NULL)%>% 
  arrange(regulatoryGene) %>% 
  dplyr::rename("Module specificity" = "weight")

write_csv(figure5supp,file.path("figures_tables", "fig5supp.csv"))

```

```{r}
#Supp figure 6 A
figure6supp_a <- ccf_aggr_all %>% 
  filter(ref_dataset=="traj" & query_dataset!="traj")%>%
  left_join(dataset_species_annot, by=c("query_dataset"="dataset"))%>%
  filter(gene %in% c("ybx1", "cebpb", "mmp9", "il1b", "amd1", "tgfbi", "txnipa")) %>% 
  mutate(species = factor(species, levels = c("mouse", "human")))

write_csv(figure6supp_a,file.path("figures_tables", "fig6supp_a.csv"))

#Supp figure 6 B,C,D
figure6supp_bcd <- map(names(gene_clusters_lst), function(module_x){
  ccf_all[gene_order_ortho,] %>%
    rownames_to_column("gene_symbol") %>%
    mutate(gene_symbol = factor(gene_symbol, levels = rev(gene_symbol))) %>% 
        arrange(gene_symbol) %>% 
    pivot_longer(-gene_symbol, values_to = "lags", names_to = "dataset") %>% 
    mutate(species = ifelse(dataset %in% c("xie","ballesteros","neutrotime"), "mouse", "human")) %>% 
    left_join(gene_clusters_lst %>% stack %>% dplyr::rename("module" = "ind"), by = c("gene_symbol"="values")) %>% 
    filter(module == module_x) %>%
    mutate(gene_symbol = factor(gene_symbol, level = unique(gene_symbol)),
          species = factor(species, levels = c("mouse","human")))
  }) %>% 
  bind_rows()

write_csv(figure6supp_bcd,file.path("figures_tables", "fig6supp_b.csv"))

#Supp figure 6 E,F,G

```

```{r}
#Supp figure 7a
immature_neut_gene_sets_zfish <- msigdbr::msigdbr(species = "Danio rerio") %>%
 filter(gs_name == "HAY_BONE_MARROW_IMMATURE_NEUTROPHIL")
figure7supp_a <- lapply(gene_clusters_lst, intersect,immature_neut_gene_sets_zfish$gene_symbol) %>%
  stack %>%
  group_by(ind, values) %>%
  arrange() %>%
  ungroup %>%
  dplyr::rename(HAY_zfish_common = "values",
                zfish_module = "ind") 

write_csv(figure7supp_a,file.path("figures_tables", "fig8supp_a.csv"))

#Supp figure 7b
immature_neut_gene_sets <- msigdbr::msigdbr(species = "Homo sapiens")%>%
 filter(gs_name == "HAY_BONE_MARROW_IMMATURE_NEUTROPHIL")

ramirez_norm_human <- ramirez_count %>%
  filter(hgnc_symbol %in% immature_neut_gene_sets$gene_symbol) %>%
  column_to_rownames("hgnc_symbol") %>% 
  mutate_all(~log2(.x+1))

ramirez_scaled_human <- t(ramirez_norm_human) %>% scale %>% t() %>% as.data.frame
figure8supp_b <- apply(ramirez_scaled_human, 2, function(x){
  ifelse(abs(x)>1.5, 1.5* x/abs(x),x)
  }) %>% 
  as.data.frame() %>% 
  rownames_to_column("gene_sympol")

write_csv(figure7supp_b,file.path("figures_tables", "fig8supp_b.csv"))
```

```{r}
#Supp figure 8 A, B
figure8_a_b <- figure7_a

write_csv(figure8_a_b,file.path("figures_tables", "fig8supp_ab.csv"))
```

```{r}
#Supp figure 16 A, B
figure16_a_b <- list(phases = enframe(phase_n, name = "Clusters number", value = "Kelley- Gardner- Sutcliffe penalty"),
          modules =enframe(phase_n, name = "Clusters number", value = "Kelley- Gardner- Sutcliffe penalty")) %>% 
  bind_rows(.id = "Cluster type")

write_csv(figure16_a_b,file.path("figures_tables", "fig16supp_ab.csv"))
```

```{r}
tabs <- c(list.files(here("figures_tables"), pattern = "fig[0-9]_",full.names = TRUE) %>% sort,
          list.files(here("figures_tables"), pattern = "fig[0-9]supp_",full.names = TRUE) %>% sort,
          list.files(here("figures_tables"), pattern = "fig_[a-z],full.names = TRUE",full.names = TRUE) %>% sort)
names(tabs) <- tabs %>% basename() %>% str_replace("\\.csv", "")
lapply(tabs, read_csv) %>% 
  openxlsx::write.xlsx(file = here("figures_tables","figures_data.xlsx"))
```

# Supplementary tables

```{r supplementary_table_1}
#supplementary_table_1: maturation-associated genes statistics
zfish_deg_summary %>%
  mutate(module_zebrafish = str_replace_all(module, "odule_", "") %>% toupper,
         module = NULL,
         module_ortho = ifelse(gene_symbol %in% gene_order_ortho,paste0(module_zebrafish, "_ortho"), NA),
         module_pan = ifelse(gene_symbol %in% unlist(consistent_specific_gene_clusters_lst),paste0(module_zebrafish, "_pan"), NA)) %>%
  dplyr::select(gene_symbol:pvalue_adj,module_zebrafish,module_ortho,module_pan) %>% 
    openxlsx::write.xlsx(here("supp_tables","supplementary_tables_1.xlsx"))
```

```{r supplementary_table_2}
#supplementary_table_2: Figure 4.A heatmap values and phase average 
zfish_deg_summary %>%
  mutate(module_zebrafish = str_replace_all(module, "odule_", "") %>% toupper,
         module = NULL,
         module_ortho = ifelse(gene_symbol %in% sc_bulk_gene_order,paste0(module_zebrafish, "_ortho"), NA),
         module_pan = ifelse(gene_symbol %in% unlist(consistent_specific_gene_clusters_lst),paste0(module_zebrafish, "_pan"), NA)) %>%
  dplyr::select(gene_symbol,phase_1_avg:traj_100) %>% 
    openxlsx::write.xlsx(here("supp_tables","supplementary_tables_2.xlsx"))
```

```{r supplementary_table_3}
#supplementary_table_3: Save hypeR msig results
hypeR_res_msig %>%
  bind_rows(.id = "Module") %>%
  as_tibble() %>%
    openxlsx::write.xlsx(here("supp_tables","supplementary_tables_3.xlsx"))
```

```{r supplementary_table_4}
#supplementary_table_4: save dtw distance and similarity
all_tf_weight_module_df %>%
  dplyr::rename(TF = "regulatoryGene",
                TF_module = "regulatory_module",
                dtw_similarity = "weight",
                dtw_distance = "dist") %>%
    openxlsx::write.xlsx(here("supp_tables","supplementary_tables_4.xlsx"))
```

```{r supplementary_table_5}
#supplementary_table_5: save dunn test statistics
reg_dunn %>%
  dplyr::select(regulatoryGene, regulatory_module, everything()) %>%
  dplyr::rename(TF = "regulatoryGene",
                TF_module = "regulatory_module") %>%
  mutate(TF_module = str_replace_all(TF_module, "odule_", "") %>% toupper,
         .y. = ifelse(str_detect(.y., "weight"), "dtw_similarity","")) %>%
    openxlsx::write.xlsx(here("supp_tables","supplementary_tables_5.xlsx"))
```

```{r supplementary_table_6}
#supplementary_table_6: Orthologous genes across single datasets

#mouse ortho in sc data
mouse_sc <- cbind(xie = homo_df_xie_1_to_1 %$% setNames(mouse_symbol, ZFIN.symbol)[unlist(consistent_specific_gene_clusters_lst)],
                  neutime = homo_df_neutime_1_to_1 %$% setNames(mouse_symbol, ZFIN.symbol)[unlist(consistent_specific_gene_clusters_lst)],
                  ballesteros = homo_df_ballesteros_1_to_1 %$% setNames(mouse_symbol, ZFIN.symbol)[unlist(consistent_specific_gene_clusters_lst)]) %>%
            as_tibble %>%
            mutate(common_gene = apply(., 1, function(x) {paste(unique(x), collapse = ",")}),
                   n = apply(., 1, function(x) {x %>% unique %>% length}))
#human ortho in sc data
human_sc <- cbind(abs = homo_df_abs_1_to_1 %$% setNames(human_symbol, ZFIN.symbol)[unlist(consistent_specific_gene_clusters_lst)],
                  suo = homo_df_suo_1_to_1 %$% setNames(human_symbol, ZFIN.symbol)[unlist(consistent_specific_gene_clusters_lst)],
                  tabula_sapiens = homo_df_tabula_sapiens_1_to_1 %$% setNames(human_symbol, ZFIN.symbol)[unlist(consistent_specific_gene_clusters_lst)]) %>%
            as_tibble %>%
            mutate(common_gene = apply(., 1, function(x) {paste(unique(x), collapse = ",")}),
                   n = apply(., 1, function(x) {x %>% unique %>% length}))

#supplementary table 6 of ortho genes 
data.frame(Zebrafish = unlist(consistent_specific_gene_clusters_lst, use.names=FALSE),
           Mouse_ortho = mouse_sc$common_gene,
           Human_ortho = human_sc$common_gene) %>%
    openxlsx::write.xlsx(here("supp_tables","supplementary_tables_6.xlsx"))
```

```{r supplementary_table_7}
#supplementary_table_7: cross-correlation results
cbind(rename_all(ccf_all, ~paste0(.x, "_max_CC")),
      rename_all(ccf_all_mean_species, ~paste0(.x, "_mean_CC"))) %>%
          round(2) %>% 
    rownames_to_column("gene_symbol") %>%
    openxlsx::write.xlsx(here("supp_tables","supplementary_tables_7.xlsx"))
```

```{r supplementary_table_9}
cell_ranger_metrics <-  Sys.glob('/home/mohamed/bioinf_isilon/Research/HALBRITTER/Internal/out/neut_zfish/results/cellranger_reporters/*/outs/metrics_summary.csv')
map_df(cell_ranger_metrics,read_csv) %>% 
  mutate(`Sequencing Run` = str_extract(cell_ranger_metrics,"MF317_A._GEX")) %>%
  dplyr::select(`Sequencing Run`, everything())%>%
    openxlsx::write.xlsx(here("supp_tables","supplementary_tables_9.xlsx"))
```

```{r supplementary_table_10}
#supplementary_table_2: Figure 4.A heatmap values and phase average 
merge_sc_bulk <- purrr::reduce(list(rownames_to_column(bulk_mean_expr_df[rownames(bulk_quant_all_scale),], "gene_symbol"),
                                    rownames_to_column(all_aggr_expr_avg_df, "gene_symbol")),
              full_join , by = "gene_symbol") %>%
              column_to_rownames("gene_symbol")
merge_sc_bulk[gene_order_ortho,] %>%
  rownames_to_column("gene_symbol") %>%
  openxlsx::write.xlsx(here("supp_tables","supplementary_tables_2.xlsx"))

```

# cross species heatmap table
```{r}
cross_species_heatmap_smoothed <- cbind(module=gene_clusters_avg_hc[gene_order_ortho,"module"],
                                 ccf_all_mean_species[gene_order_ortho,] %>% as.data.frame %>% rename_with(~paste0("cc_mean_", .x)),
                                 (ls_aggr_expr_smooth%>%
                                   do.call(cbind,.))[gene_order_ortho,]) %>%
  rownames_to_column("gene_symbol") %>% 
  as.data.frame()
write_csv(cross_species_heatmap_smoothed,"cross_species_heatmap_smoothed_1500.csv")
```

# stage-aggregated heatmap table
```{r}
norm_lst$quants_all %>%
  as.data.frame() %>%
  rownames_to_column("gene_symbol") %>%
  mutate(module=gene_clusters_avg_hc[gene_symbol,"module"]) %>%
  dplyr::select(gene_symbol, module, everything()) %>%
  write_csv("scaled_stage_aggr_quantile.csv")

``` 

# Merge all tables in a multi-sheet excel file
```{r supplementary_tables_all}
supp_tables <- list.files(here("supp_tables/"), full.names=TRUE)
lapply(supp_tables, openxlsx::read.xlsx) %>%
 setNames(.,supp_tables%>%str_extract("supplementary_table_.")) %>%
    openxlsx::write.xlsx(here("supp_tables","supplementary_tables_all.xlsx"))
```